source/Nevermore/Advanced/Concurrency/ITransactionConcurrencyHandler.cs

```diff
@@ -0,0 +1,13 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Nevermore.Advanced.Concurrency
+{
+    public interface ITransactionConcurrencyHandler : IDisposable
+    {
+        IDisposable Lock();
+        
+        Task<IDisposable> LockAsync(CancellationToken cancellationToken);
+    }
```

Comment:
We decided to replace the `DeadlockAwareLock` with this abstraction as a way for us to enable logging without locking. It's a bit much, but felt cleaner than introducing a bunch of additional checks into a single class.
Open to alternatives.

Reply:
Here's a spike of what this would look like if we had a single type: https://github.com/OctopusDeploy/Nevermore/pull/289